community/tutorials/gdnative/gdnative-c-example.rst

```diff
@@ -0,0 +1,184 @@
+Note that in GDNative we do not (yet) tell Godot what parameters it needs to provide when calling this method. Godot will simply give us an array of variants for each parameter that was provided. We'll have to do sufficient checking when executing our method.
+
+.. code:: C
+
+    void GDN_EXPORT godot_nativescript_init(void *p_handle) {
+        godot_instance_create_func create = { NULL, NULL, NULL };
+        create.create_func = &simple_constructor;
+
+        godot_instance_destroy_func destroy = { NULL, NULL, NULL };
+        destroy.destroy_func = &simple_destructor;
+
+        nativescript_api->godot_nativescript_register_class(p_handle, "SIMPLE", "Reference", create, destroy);
+
+        godot_instance_method get_data = { NULL, NULL, NULL };
+        get_data.method = &simple_get_data;
+
+        godot_method_attributes attributes = { GODOT_METHOD_RPC_MODE_DISABLED };
+
+        nativescript_api->godot_nativescript_register_method(p_handle, "SIMPLE", "get_data", attributes, get_data);
+    }
+
+Now it's time to start working on the functions of our object. First we define a structure that we use to store the member data of an instance of our GDNative class. 
+
+.. code:: C
+
+    typedef struct user_data_struct {
+        char data[256];
+    } user_data_struct;
+
+And then we define our constructor. All we do in our constructor is allocate memory for our structure and fill it with some data. Note that we use godots memory functions so the memory gets tracked and then return the pointer to our new structure. This pointer will act as our instance identifier in case multiple objects are instantiated. 
+
+.. code:: C
+
+    GDCALLINGCONV void *simple_constructor(godot_object *p_instance, void *p_method_data) {        
+        user_data_struct *user_data = api->godot_alloc(sizeof(user_data_struct));
+        strcpy(user_data->data, "World from GDNative!");
+
+        return user_data;
+    }
+
+Our destructor is called when Godot is done with our object and we offcourse free our instance data. Note that p_user_data will always point to the structure we allocated in our constructor.
+
+.. code:: C
+
+    GDCALLINGCONV void simple_destructor(godot_object *p_instance, void *p_method_data, void *p_user_data) {
+        api->godot_free(p_user_data);
+    }
+
+And finally we implement our get_data function. Data is always sent and returned as variants so in order to return our data, which is a string, we first need to convert our C string to a Godot string object, and then copy that string object into the variant we are returning. Don't forget to destroy the string data or we'll leak memory! The variant we return is destroyed automatically by Godot.
```

Comment:
The "Don't forget to destroy the string data or we'll leak memory!" should be explained a bit further.

"
Strings are heap-allocated in Godot, so they have a destructor which frees the memory. Destructors are named `godot_TYPENAME_destroy`. When a Variant gets created with a String it references the String. That means that the original String can be "destroyed" to decrease the ref-count. If that does not happen the String memory will leak since the ref-count will never be zero and the memory never deallocated. The returned variant gets automatically destroyed by Godot.

(In more complex operations it can be confusing the keep track of which value needs to be deallocated and which does not. As a general rule: call `godot_XXX_destroy` when a C++ destructor would be called instead. The String destructor would be called in C++  after the Variant was created, so the same is necessary in C)
"